### ISSUE TITLE
pbspro controller start fails

### ISSUE THREAD

**Issue by razZ0r** [chapmanb/bcbio-nextgen#1276]
I'm trying to run bcbio on a cluster with PBSPro. Among many problems I have, the first is that the controller script fails.

Startup command is:

<pre>
/pico/home/userexternal/esebesty/work/bcbio-20160314/bcbio/tool/bin/bcbio_nextgen.py ../config/single-pair-cineca-test-resources.yaml ../config/single-pair-cineca-test-interactive-samples.yaml -t ipython -s pbspro -q parallel -n 16 -r account=try16_EMUTPRED
</pre>

I have some output, etc and then the command is stuck, and 0 engines are running.

In the work/log/ipython/log dir, the log says:

<pre>
2016-03-16 19:18:03.953 [IPClusterStart] Starting ipcluster with [daemon=True]
2016-03-16 19:18:03.959 [IPClusterStart] Creating pid file: /pico/scratch/userexternal/esebesty/bcbio_run/single-pair-cineca-interactive/work/log/ipython/pid/ipcluster-e71c8d9d-02d5-42f5-ae0c-a1290b5bc5b0.pid
2016-03-16 19:18:03.960 [IPClusterStart] Starting Controller with cluster_helper.cluster.BcbioPBSPROControllerLauncher
2016-03-16 19:18:03.960 [IPClusterStart] Starting BcbioPBSPROControllerLauncher: ['qsub', u'./pbspro_controller']
2016-03-16 19:18:03.960 [IPClusterStart] adding PBS queue settings to batch script
2016-03-16 19:18:03.960 [IPClusterStart] adding job array settings to batch script
2016-03-16 19:18:03.961 [IPClusterStart] Writing batch script: ./pbspro_controller
2016-03-16 19:18:03.968 [IPClusterStart] ERROR | Controller start failed
Traceback (most recent call last):
  File "/pico/home/userexternal/esebesty/work/bcbio-20160314/bcbio/data/anaconda/lib/python2.7/site-packages/ipyparallel/apps/ipclusterapp.py", line 506, in start_controller
    self.controller_launcher.start()
  File "/pico/home/userexternal/esebesty/work/bcbio-20160314/bcbio/data/anaconda/lib/python2.7/site-packages/cluster_helper/cluster.py", line 718, in start
    return super(BcbioPBSPROControllerLauncher, self).start(1)
  File "/pico/home/userexternal/esebesty/work/bcbio-20160314/bcbio/data/anaconda/lib/python2.7/site-packages/ipyparallel/apps/launcher.py", line 1151, in start
    output = check_output(self.args, env=os.environ)
  File "/pico/home/userexternal/esebesty/work/bcbio-20160314/bcbio/data/anaconda/lib/python2.7/subprocess.py", line 566, in check_output
    process = Popen(stdout=PIPE, *popenargs, **kwargs)
  File "/pico/home/userexternal/esebesty/work/bcbio-20160314/bcbio/data/anaconda/lib/python2.7/subprocess.py", line 710, in __init__
    errread, errwrite)
  File "/pico/home/userexternal/esebesty/work/bcbio-20160314/bcbio/data/anaconda/lib/python2.7/subprocess.py", line 1335, in _execute_child
    raise child_exception
OSError: [Errno 2] No such file or directory
</pre>

The pbspro_controller script is there, so I'm wondering what might be the source of the "No such file or directory" error.

<pre>
#!/bin/sh

#PBS -q parallel
#PBS -V
#PBS -N bcbio-c
#PBS -l select=1:ncpus=1
cd $PBS_O_WORKDIR
/pico/home/userexternal/esebesty/work/bcbio-20160314/bcbio/data/anaconda/bin/python -E -c 'import resource; cur_proc, max_proc = resource.getrlimit(resource.RLIMIT_NPROC); target_proc = min(max_proc, 10240) if max_proc > 0 else 10240; resource.setrlimit(resource.RLIMIT_NPROC, (max(cur_proc, target_proc), max_proc)); cur_hdls, max_hdls = resource.getrlimit(resource.RLIMIT_NOFILE); target_hdls = min(max_hdls, 10240) if max_hdls > 0 else 10240; resource.setrlimit(resource.RLIMIT_NOFILE, (max(cur_hdls, target_hdls), max_hdls)); from cluster_helper.cluster import VMFixIPControllerApp; VMFixIPControllerApp.launch_instance()' --ip=* --log-to-file --profile-dir="/pico/scratch/userexternal/esebesty/bcbio_run/single-pair-cineca-interactive/work/log/ipython" --cluster-id="e71c8d9d-02d5-42f5-ae0c-a1290b5bc5b0" --nodb --hwm=1 --scheme=leastload --HeartMonitor.max_heartmonitor_misses=720 --HeartMonitor.period=5000
</pre>

**Comment by chapmanb** [comment]
Endre;
Sorry about the issues. PBSPro is definitely one of our less tested integrations so happy to try and fix issues.

This failure looks like your missing qsub from your PATH. bcbio will create the controller script, then try to submit it with `qsub pbspro_controller`. Does adding qsub in help things work cleanly?

**Comment by razZ0r** [comment]
No, qsub is in my path. Maybe bcbio is not running the command from the work/ directory, so referring to ./pbspro_controller fails?

**Comment by roryk** [comment]
Does running `qsub pbspro_controller` and `qsub pbspro_engines` spin up a controller and engines?

**Comment by razZ0r** [comment]
No, its missing the account number. 

<pre>
qsub: ERROR: Account number None is not valid for user esebesty
</pre>

And the line `#PBS -A try16_EMUTPRED` is missing from the `pbspro_controller` script, even if I started bcbio with `-r account=try16_EMUTPRED`.  Can this lead to the `OSError: [Errno 2] No such file or directory` error?


No `pbspro_engines` is written anywhere in the work directory.

**Comment by roryk** [comment]
Thanks, the two issues might be unrelated but both will have to be solved to get it to work. If you add the missing line to the controller script, does it start up?

**Comment by razZ0r** [comment]
Should have tried that already. Yes, if I add the line the controller starts.

**Comment by roryk** [comment]
Great-- do you think you could clone this repository:

https://github.com/roryk/ipython-cluster-helper

and install it with the python that bcbio is using:
```bash
/pico/home/userexternal/esebesty/work/bcbio-20160314/bcbio/data/anaconda/bin/python setup.py install
```

and then run the example script:

```bash
/pico/home/userexternal/esebesty/work/bcbio-20160314/bcbio/data/anaconda/bin/python example/example.py --scheduler pbspro --num_jobs 3 --resources account=try16_EMUTPRED --queue parallel
```

and see if it fails the same way? That way we can solve the problems there.

**Comment by razZ0r** [comment]
OK, installed it and ran the script:

<pre>
2016-03-17 16:45:50.492 [IPClusterStart] Starting ipcluster with [daemon=True]
2016-03-17 16:45:50.498 [IPClusterStart] Creating pid file: /pico/home/userexternal/esebesty/.ipython/profile_51e5d3ce-ec57-11e5-a77b-6cae8b08c5c3/pid/ipcluster-6f2c614e-8f55-4827-9beb-9ac82b57effb.pid
2016-03-17 16:45:50.499 [IPClusterStart] Starting Controller with cluster_helper.cluster.BcbioPBSPROControllerLauncher
2016-03-17 16:45:50.499 [IPClusterStart] Starting BcbioPBSPROControllerLauncher: ['qsub', u'./pbspro_controller']
2016-03-17 16:45:50.499 [IPClusterStart] adding PBS queue settings to batch script
2016-03-17 16:45:50.499 [IPClusterStart] adding job array settings to batch script
2016-03-17 16:45:50.500 [IPClusterStart] Writing batch script: ./pbspro_controller
2016-03-17 16:45:50.584 [IPClusterStart] ERROR | Controller start failed
Traceback (most recent call last):
  File "/pico/home/userexternal/esebesty/work/bcbio-20160314/bcbio/data/anaconda/lib/python2.7/site-packages/ipyparallel/apps/ipclusterapp.py", line 506, in start_controller
    self.controller_launcher.start()
  File "cluster_helper/cluster.py", line 718, in start
    return super(BcbioPBSPROControllerLauncher, self).start(1)
  File "/pico/home/userexternal/esebesty/work/bcbio-20160314/bcbio/data/anaconda/lib/python2.7/site-packages/ipyparallel/apps/launcher.py", line 1151, in start
    output = check_output(self.args, env=os.environ)
  File "/pico/home/userexternal/esebesty/work/bcbio-20160314/bcbio/data/anaconda/lib/python2.7/subprocess.py", line 573, in check_output
    raise CalledProcessError(retcode, cmd, output=output)
CalledProcessError: Command '['qsub', u'./pbspro_controller']' returned non-zero exit status 32
</pre>

**Comment by roryk** [comment]
Great, I opened a branch over there so we can work on a fix. If you:

```
git pull
```

and then switch to the branch:

```
git checkout pbspro-fix
```

and run the setup.py again, does the account get added to the script?

**Comment by razZ0r** [comment]
I have a line in the output with the account added:

<pre>
2016-03-17 17:00:23.191 [IPClusterStart] {'BaseIPythonApplication': {'profile': u'5a19c9e0-ec59-11e5-a31e-6cae8b08c5c3'}, 'BcbioPBSPROControllerLauncher': {'mem': '', 'tag': '', 'cores': 1, 'resources': 'account=try16_EMUTPRED'}, 'BcbioPBSPROEngineSetLauncher': {'mem': '1', 'tag': '', 'cores': 1, 'resources': 'account=try16_EMUTPRED'}, 'IPClusterEngines': {'early_shutdown': 240}, 'PBSPROLauncher': {'queue': 'parallel'}, 'Application': {'log_level': 10}, 'BaseParallelApplication': {'log_to_file': True, 'cluster_id': u'c8003e01-26d5-425a-945a-e024a64b0ad1'}, 'IPClusterStart': {'controller_launcher_class': u'cluster_helper.cluster.BcbioPBSPROControllerLauncher', 'engine_launcher_class': u'cluster_helper.cluster.BcbioPBSPROEngineSetLauncher', 'n': 3, 'delay': 10, 'daemonize': True}}
</pre>

but in the logfile:

<pre>
2016-03-17 17:00:23.331 [IPClusterStart] Starting ipcluster with [daemon=True]
2016-03-17 17:00:23.338 [IPClusterStart] Creating pid file: /pico/home/userexternal/esebesty/.ipython/profile_5a19c9e0-ec59-11e5-a31e-6cae8b08c5c3/pid/ipcluster-c8003e01-26d5-425a-945a-e024a64b0ad1.pid
2016-03-17 17:00:23.338 [IPClusterStart] Starting Controller with cluster_helper.cluster.BcbioPBSPROControllerLauncher
2016-03-17 17:00:23.339 [IPClusterStart] ERROR | Controller start failed
Traceback (most recent call last):
  File "/pico/home/userexternal/esebesty/work/bcbio-20160314/bcbio/data/anaconda/lib/python2.7/site-packages/ipyparallel/apps/ipclusterapp.py", line 506, in start_controller
    self.controller_launcher.start()
  File "cluster_helper/cluster.py", line 719, in start
    resources = "\n".join(_prep_pbspro_resources(self.resources))
AttributeError: 'BcbioPBSPROControllerLauncher' object has no attribute 'resources'
</pre>

and no new pbspro_controller is written anywhere.

**Comment by roryk** [comment]
Thanks, sorry-- it is tough not being able to test. I pushed a fix for that.

**Comment by razZ0r** [comment]
No worries, at least I can help testing! Similar error as in https://github.com/chapmanb/bcbio-nextgen/issues/1276#issuecomment-197942724

<pre>
2016-03-17 17:18:48.443 [IPClusterStart] Starting ipcluster with [daemon=True]
2016-03-17 17:18:48.452 [IPClusterStart] Creating pid file: /pico/home/userexternal/esebesty/.ipython/profile_ed091f9c-ec5b-11e5-9d6b-6cae8b08c5c3/pid/ipcluster-48c1ba07-d731-4794-91c6-acd916f87637.pid
2016-03-17 17:18:48.453 [IPClusterStart] Starting Controller with cluster_helper.cluster.BcbioPBSPROControllerLauncher
2016-03-17 17:18:48.454 [IPClusterStart] Starting BcbioPBSPROControllerLauncher: ['qsub', u'./pbspro_controller']
2016-03-17 17:18:48.455 [IPClusterStart] adding PBS queue settings to batch script
2016-03-17 17:18:48.455 [IPClusterStart] adding job array settings to batch script
2016-03-17 17:18:48.456 [IPClusterStart] Writing batch script: ./pbspro_controller
2016-03-17 17:18:48.570 [IPClusterStart] ERROR | Controller start failed
Traceback (most recent call last):
  File "/pico/home/userexternal/esebesty/work/bcbio-20160314/bcbio/data/anaconda/lib/python2.7/site-packages/ipyparallel/apps/ipclusterapp.py", line 506, in start_controller
    self.controller_launcher.start()
  File "cluster_helper/cluster.py", line 725, in start
    return super(BcbioPBSPROControllerLauncher, self).start(1)
  File "/pico/home/userexternal/esebesty/work/bcbio-20160314/bcbio/data/anaconda/lib/python2.7/site-packages/ipyparallel/apps/launcher.py", line 1151, in start
    output = check_output(self.args, env=os.environ)
  File "/pico/home/userexternal/esebesty/work/bcbio-20160314/bcbio/data/anaconda/lib/python2.7/subprocess.py", line 573, in check_output
    raise CalledProcessError(retcode, cmd, output=output)
CalledProcessError: Command '['qsub', u'./pbspro_controller']' returned non-zero exit status 32
</pre>

and the pbspro_controller still missing the `#PBS -A try16_EMUTPRED` line.

**Comment by roryk** [comment]
Hmm-- when I run it I get it in there:

```
python example/example.py --scheduler pbspro --num_jobs 3 --queue parallel --resources account=foobar
```

```
#!/bin/sh

#PBS -q parallel
#PBS -V
#PBS -N bcbio-c
#PBS -l select=1:ncpus=1
#PBS -A foobar
#PBS -l walltime=239:00:00
cd $PBS_O_WORKDIR
/Users/rory/.conda/bin/python -E -c 'import resource; cur_proc, max_proc = resource.getrlimit(resource.RLIMIT_NPROC); target_proc = min(max_proc, 10240) if max_proc > 0 else 10240; resource.setrlimit(resource.RLIMIT_NPROC, (max(cur_proc, target_proc), max_proc)); cur_hdls, max_hdls = resource.getrlimit(resource.RLIMIT_NOFILE); target_hdls = min(max_hdls, 10240) if max_hdls > 0 else 10240; resource.setrlimit(resource.RLIMIT_NOFILE, (max(cur_hdls, target_hdls), max_hdls)); from cluster_helper.cluster import VMFixIPControllerApp; VMFixIPControllerApp.launch_instance()' --ip=* --log-to-file --profile-dir="/Volumes/Clotho/Users/rory/.ipython/profile_c3ed675c-ec5c-11e5-823c-406c8f3b7df5" --cluster-id="b7633e30-d075-4a18-9457-aff8e0030f65" --nodb --hwm=1 --scheme=leastload --HeartMonitor.max_heartmonitor_misses=720 --HeartMonitor.period=5000
```

**Comment by razZ0r** [comment]
I have to change the default walltime, as it is over the max allowed. If I write in the command line only `--resources account=try16_EMUTPRED` the -A option is there, if I specify `--resources account=try16_EMUTPRED,walltime=01:00:00` or `--resources account=try16_EMUTPRED --resources walltime=01:00:00` it is missing.

**Comment by roryk** [comment]
Great-- what about separating them by a ; instead of a ,?

**Comment by razZ0r** [comment]
Success! The controller is running. Have to write `--resources 'account=try16_EMUTPRED;walltime=01:00:00'` though, with single quotes. Otherwise I just get:

<pre>
Please specify --local to run locally or a scheduler and queueto run on with --scheduler and --queue
-bash: --queue: command not found
</pre>

**Comment by roryk** [comment]
Gotcha-- do the engines never come up? Is there an engine file now?

**Comment by razZ0r** [comment]
There is an engine file, but the engines don't start. Now I have an error in ipcluster log:

<pre>
2016-03-17 17:48:48.211 [IPClusterStart] Starting ipcluster with [daemon=True]
2016-03-17 17:48:48.220 [IPClusterStart] Creating pid file: /pico/home/userexternal/esebesty/.ipython/profile_1dcac924-ec60-11e5-afd5-6cae8b08c5c3/pid/ipcluster-815d3e84-b173-4f54-a305-0f9fe664457d.pid
2016-03-17 17:48:48.221 [IPClusterStart] Starting Controller with cluster_helper.cluster.BcbioPBSPROControllerLauncher
2016-03-17 17:48:48.221 [IPClusterStart] Starting BcbioPBSPROControllerLauncher: ['qsub', u'./pbspro_controller']
2016-03-17 17:48:48.221 [IPClusterStart] adding PBS queue settings to batch script
2016-03-17 17:48:48.221 [IPClusterStart] adding job array settings to batch script
2016-03-17 17:48:48.222 [IPClusterStart] Writing batch script: ./pbspro_controller
2016-03-17 17:48:48.434 [IPClusterStart] Job submitted with job id: u'259763'
2016-03-17 17:48:48.434 [IPClusterStart] Process 'qsub' started: u'259763'
2016-03-17 17:48:58.437 [IPClusterStart] Starting 3 Engines with cluster_helper.cluster.BcbioPBSPROEngineSetLauncher
2016-03-17 17:48:58.437 [IPClusterStart] Starting BcbioPBSPROEngineSetLauncher: ['qsub', u'./pbspro_engines']
2016-03-17 17:48:58.438 [IPClusterStart] adding PBS queue settings to batch script
2016-03-17 17:48:58.438 [IPClusterStart] adding job array settings to batch script
2016-03-17 17:48:58.438 [IPClusterStart] Writing batch script: ./pbspro_engines
2016-03-17 17:48:58.450 [IPClusterStart] ERROR | Engine start failed
Traceback (most recent call last):
  File "/pico/home/userexternal/esebesty/work/bcbio-20160314/bcbio/data/anaconda/lib/python2.7/site-packages/ipyparallel/apps/ipclusterapp.py", line 331, in start_engines
    self.engine_launcher.start(self.n)
  File "cluster_helper/cluster.py", line 698, in start
    return super(BcbioPBSPROEngineSetLauncher, self).start(n)
  File "/pico/home/userexternal/esebesty/work/bcbio-20160314/bcbio/data/anaconda/lib/python2.7/site-packages/ipyparallel/apps/launcher.py", line 1151, in start
    output = check_output(self.args, env=os.environ)
  File "/pico/home/userexternal/esebesty/work/bcbio-20160314/bcbio/data/anaconda/lib/python2.7/subprocess.py", line 573, in check_output
    raise CalledProcessError(retcode, cmd, output=output)
CalledProcessError: Command '['qsub', u'./pbspro_engines']' returned non-zero exit status 255
</pre>

**Comment by roryk** [comment]
All right, one step closer. If you submit the engine file manually (qsub pbspro_engines) does it give a more useful error?

**Comment by razZ0r** [comment]
I guess so:

<pre>
qsub:  cannot submit non-rerunable Array Job
qsub:  cannot submit non-rerunable Array Job
qsub: directive error: -r n -W umask
</pre>

**Comment by razZ0r** [comment]
This might be specific to the cluster. See http://www.hpc.cineca.it/content/batch-scheduler-pbs and the Array jobs section.

**Comment by roryk** [comment]
That stinks-- maybe pinging the admins to see if you can get an exception would be helpful.

How do you specify jobs and cores with PBSPro that aren't an array? Let's say 3 jobs with 8 cores each. If you set that in the script and take out the job array section, do the engines run? We could maybe just not submit job arrays.

**Comment by razZ0r** [comment]
I'll ask around. I just started using this cluster this week and never used pbspro before, so no idea.

**Comment by roryk** [comment]
Ok, let us know what you find out. Thanks for helping debug.

**Comment by razZ0r** [comment]
One question: is it important the jobs are non rerunable? Why?

**Comment by razZ0r** [comment]
If I submit the pbspr_engine, with directives like this:

<pre>
#PBS -q parallel
#PBS -V
#PBS -N bcbio-e
#PBS -l select=3:ncpus=1:mem=1024mb
#PBS -A try16_EMUTPRED
#PBS -l walltime=01:00:00
</pre>

it dies with:

<pre>
2016-03-18 10:37:56.034 [IPEngineApp] CRITICAL | Profile directory u'/pico/home/userexternal/esebesty/.ipython/profile_1dcac924-ec60-11e5-afd5-6cae8b08c5c3' not found.
</pre>

If I create the directory manually, the qsub job is submitted, running, but nothing gets written to the profile dir.

**Comment by razZ0r** [comment]
I was told to just submit a job with select=1:ncpus=8 three times. Each submission will have its own job ID and will be independent from the others. Is this feasible?

**Comment by chapmanb** [comment]
Endre;
Thanks for all the patience debugging this. All of our support for PBSPro has been via back and forth like this since we don't run those systems locally so appreciate the help. Regarding you last three posts:

- I don't believe the non-rerunable flag is important. That might be leftover from someone's suggestion but Rory would know for sure.
- Did you also start the controller for that engine job that died? It looks like it's complaining because it can't find the directory that contains details about connecting to the controller. If you're only testing submission, that's expected since there is no controller.
- Not using job arrays would be a large change in how we submit and manage jobs. Right now we let the queuing software take care of managing multiple engines, so would have to re-write all that logic in bcbio. What concern do you have about allowing job arrays?

Hope this helps.

**Comment by razZ0r** [comment]
OK, so now the account, walltime, etc is passed to the pbspro_controller and pbspro_engines scripts. 

The engines script fails, because the #PBS -r y directive is missing from it. I did not get an answer on why cineca has this restriction, but I guess this is not the case everywhere.

If the non-rerunable flag is not important, maybe the best thing would be to write explicitly the #PBS -r y directive to the engine script. Or I could specify it on the command line, together with other resources options, so the whole thing is more flexible.

I'm very motivated to make the pipeline run on this cluster, as I have a 2x overall speed increase, with half the cpu cores. :)

**Comment by roryk** [comment]
I think the rerunnable flag is important to be `n` because it tells the scheduler that it is not okay to kill this job and resubmit it.

**Comment by razZ0r** [comment]
Hmmmm, I'll try to convince the admins to allow us using the non-rerunnable flag.

**Comment by roryk** [comment]
Sorry about that, this is the part of the parallel setup that is brittle. The engines have to remain up, they act like a conduit for the jobs to get executed on the machines they are on. There isn't a mechanism implemented to detect and resubmit engine jobs that are killed, so once they are killed everything falls apart.

**Comment by razZ0r** [comment]
Hi, it turned out that this is not a local setting at the cluster I'm using, but a PBSPro restriction. The PBSPro [manual](http://www.hpc.cineca.it/sites/default/files/PBSProUserGuide13.0beta.pdf) says:

> Job arrays are required to be rerunnable. PBS will not accept a job array that is marked as
not rerunnable. You can submit a job array without specifying whether it is rerunnable,
and PBS will automatically mark it as rerunnable.

No idea why this is the case. So there is not a way to support bcbio with PBSPro, unless there is some mechanism for handling rerunnable array jobs. :(

I'll probably write some script that generates a qsub job for my sample pairs, where I'm running bcbio with `-t local -n 8/16/whatever` options.

**Comment by roryk** [comment]
That sucks, I didn't know PBSPro had that limitation. Could you try the `pbspro-fix` branch on `ipython-cluster-helper` and see if the example script runs? I took out the job array and set the number of chunks in the resources section instead, similar to what you did above.

**Comment by razZ0r** [comment]
Seems to work.

<pre>
2016-03-24 10:52:01.283 [IPClusterStart] Looking for ipcluster_config in /pico/home/userexternal/esebesty/repos/misc_tools/ipython-cluster-helper
2016-03-24 10:52:01.283 [IPClusterStart] Config changed:
2016-03-24 10:52:01.284 [IPClusterStart] {'BaseIPythonApplication': {'profile': u'0d24cff2-f1a6-11e5-b5c2-6cae8b08af7d'}, 'BcbioPBSPROControllerLauncher': {'mem': '', 'tag': '', 'cores': 1, 'resources': 'account=try16_EMUTPRED;walltime=01:00:00'}, 'BcbioPBSPROEngineSetLauncher': {'mem': '1', 'tag': '', 'cores': 1, 'resources': 'account=try16_EMUTPRED;walltime=01:00:00'}, 'IPClusterEngines': {'early_shutdown': 240}, 'PBSPROLauncher': {'queue': 'parallel'}, 'Application': {'log_level': 10}, 'BaseParallelApplication': {'log_to_file': True, 'cluster_id': u'0568f3a7-2169-419f-8d8f-f43718a18fd3'}, 'IPClusterStart': {'controller_launcher_class': u'cluster_helper.cluster.BcbioPBSPROControllerLauncher', 'engine_launcher_class': u'cluster_helper.cluster.BcbioPBSPROEngineSetLauncher', 'n': 3, 'delay': 10, 'daemonize': True}}
1 Engines running
First check to see if we can talk to the engines.
This long computation that waits for 5 seconds before returning takes a while to run serially..
[('node013', 60), ('node013', 63), ('node013', 66), ('node013', 69), ('node013', 72), ('node013', 75), ('node013', 78), ('node013', 81), ('node013', 84), ('node013', 87), ('node013', 90), ('node013', 93), ('node013', 96), ('node013', 99), ('node013', 102), ('node013', 105), ('node013', 108), ('node013', 111), ('node013', 114), ('node013', 117)]
That took 20.0202670097 seconds.
Running it in parallel goes much faster...
[('node003', 60), ('node003', 63), ('node003', 66), ('node003', 69), ('node003', 72), ('node003', 75), ('node003', 78), ('node003', 81), ('node003', 84), ('node003', 87), ('node003', 90), ('node003', 93), ('node003', 96), ('node003', 99), ('node003', 102), ('node003', 105), ('node003', 108), ('node003', 111), ('node003', 114), ('node003', 117)]
That took 20.1146450043 seconds.
Sending a shutdown signal to the controller and engines.
2016-03-24 10:53:08.848 [IPClusterStop] Stopping cluster [pid=20918] with [signal=2]
</pre>

**Comment by roryk** [comment]
Ah, doesn't look like it is working since it isn't actually any faster and it is just spinning up one engine.

**Comment by ohofmann** [comment]
Currently trying to retrace this on another PBSPro system to provide additional data points. Stuck at https://github.com/chapmanb/bcbio-nextgen/issues/1276#issuecomment-197942724 -- on my side it just times out waiting for engines to come up, and I am not seeing any nodes being requested via `qstat`. I am using the main branch of https://github.com/roryk/ipython-cluster-helper (not the pbspro-fix):

```
/work/d118/d118/ohofmann/bcbio/anaconda/bin/python example/example.py --scheduler pbspro --num_jobs 3 --resources account=d118 --queue short --timeout=60

(...)

2016-03-24 13:12:19.013 [IPClusterStart] Attempting to load config file: ipcluster_d01b4f1a_aa50_4d33_9119_7299e4409461_config.py
2016-03-24 13:12:19.013 [IPClusterStart] Looking for ipengine_config in /etc/ipython
2016-03-24 13:12:19.013 [IPClusterStart] Looking for ipengine_config in /usr/local/etc/ipython
2016-03-24 13:12:19.013 [IPClusterStart] Looking for ipengine_config in /home3/d118/d118/ohofmann/.ipython/profile_082b550e-f1c2-11e5-8da3-90b11c43384c
2016-03-24 13:12:19.014 [IPClusterStart] Loaded config file: /home3/d118/d118/ohofmann/.ipython/profile_082b550e-f1c2-11e5-8da3-90b11c43384c/ipengine_config.py
2016-03-24 13:12:19.014 [IPClusterStart] Looking for ipengine_config in /fs3/d118/d118/ohofmann/ipython-cluster-helper
2016-03-24 13:12:19.014 [IPClusterStart] Attempting to load config file: ipcluster_d01b4f1a_aa50_4d33_9119_7299e4409461_config.py
2016-03-24 13:12:19.014 [IPClusterStart] Looking for ipcluster_config in /etc/ipython
2016-03-24 13:12:19.014 [IPClusterStart] Looking for ipcluster_config in /usr/local/etc/ipython
2016-03-24 13:12:19.014 [IPClusterStart] Looking for ipcluster_config in /home3/d118/d118/ohofmann/.ipython/profile_082b550e-f1c2-11e5-8da3-90b11c43384c
2016-03-24 13:12:19.015 [IPClusterStart] Loaded config file: /home3/d118/d118/ohofmann/.ipython/profile_082b550e-f1c2-11e5-8da3-90b11c43384c/ipcluster_config.py
2016-03-24 13:12:19.015 [IPClusterStart] Looking for ipcluster_config in /fs3/d118/d118/ohofmann/ipython-cluster-helper
0 Engines running
```

**Comment by lpantano** [comment]
Hi all

I will close this because it seems the discussion is staled. 

Please, open a new issue if there is still a problem.

thanks!

**Comment by roryk** [comment]
Hi @diallobakary4,

I think this issue should be fixed. Could you post your command line for bcbio-nextgen and one of the engine files that isn't working?

**Comment by diallobakary4** [comment]
I have changed the parameter -r from n to y.  The job looks like it is running as I don't get the ""qsub: cannot submit non-rerunable Array Job" error anymore.
But there is still a problem. My program is just reading one file of my array job. 
I think I'm not defining this variable for my file well: file=`ls *.com | head -$PBS_ARRAYID | tail -1
 don't exactly how the $PBS_ARRAYID  array variable evolves exactly to correctly assign the different files according to its values.

**Comment by roryk** [comment]
Hi Bakary,

Sorry about that.  We're not actually submitting job arrays in bcbio for PBSpro anymore. You can see here:

https://github.com/roryk/ipython-cluster-helper/blob/f0a86ffbcdd246ab9e768ad4df921da9a3b84864/cluster_helper/cluster.py#L695

we break it up and submit one separate batch script per job, so I'm confused where this error is coming from. Could you pass along the batch script you are submitting to bcbio and an example of one of the engine job batch scripts that are churned out so we can see what the issue is?

**Comment by diallobakary4** [comment]
Sorry if I'm confusing you. It's true that I'm mixing up everything a bit. I'm not using bcbio. I'm submitting job arrays on another cluster.  Here is the batch script I'm using. The problem is in the 
`file=`ls *.com | head -$PBS_ARRAYID | tail -1``
This is supposed to pass all the com files so that they will be all submitted. 
I have found the piece of code here https://pace.gatech.edu/submitting-job-arrays. There is a "n" that I don't understand.

`#!/bin/bash
#PBS -P 
#PBS -N 49860205
#PBS -l select=1:ncpus=24
#PBS -l place=excl
#PBS -l walltime=96:00:00
#PBS -q smp
#PBS -m be
#PBS -M fsd@fds.com
#PBS -r y
#PBS -o /mnt/lustre/users/hjh/MDs/Library_opt/output
#PBS -e /mnt/lustre/users/jh/MDs/Library_opt/error						
#PBS -J 1-9

function cleanup_scratch() {
echo "Deleting inside signal handler, meaning I probably either hit the walltime, or deleted the job using qdel"

rm -rfv $GAUSS_SCRDIR
echo "---"
echo "Signal handler ending time:"
date
exit 0
}

trap 'cleanup_scratch' SIGTERM SIGHUP SIGINT SIGQUIT SIGFPE SIGKILL SIGALRM

mkdir -p /mnt/lustre/users/hj/MDs/Library_opt/$PBS_JOBID
cd /mnt/lustre/users/bdiallo/MDs/Library_opt

echo "Job starts here"
date

module add chpc/gaussian09/D01
GAUSS_SCRDIR=/mnt/lustre/users/hj/MDs/Library_opt/$PBS_JOBID
export g09root GAUSS_SCRDIR
source $g09root/g09/bsd/g09.profile

file=`ls *.com | head -$PBS_ARRAYID | tail -1`
echo "$file"
LINDA=`cat $PBS_NODEFILE | uniq | tr '\n' , | sed 's|,$||'`
sed -i 's/.*linda.*/\%lindaworkers='$LINDA'/' "$file"

g09 < "$file" > "${file::-4}".log

echo "rm -rfv $GAUSS_SCRDIR"
pushd /mnt/lustre/users/jhj/MDs/Library_opt
rm -rf $PBS_JOBID;
popd

echo "Normal execution ending time:"
date`

Thanks you for your help.

**Comment by roryk** [comment]
Sorry Bakary, this is just for issues related to bcbio. I'd recommend asking the sysadmins running your cluster for help for this one, they'll definitely be better able to help.

**Comment by diallobakary4** [comment]
Oh Ok Roryk.
Thank you for your help though.
